spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/HttpClientBeanPostProcessor.java

```diff
@@ -259,34 +234,68 @@ protected void handle(HttpClientResponse httpClientResponse,
 			if (reference == null || reference.get() == null) {
 				return;
 			}
-			handler().handleReceive(httpClientResponse, throwable,
-					(Span) reference.get());
+			handler().handleReceive(new WrappedHttpClientResponse(httpClientResponse),
+					throwable, (Span) reference.get());
 		}
 
 	}
 
-	private static class HttpAdapter
-			extends brave.http.HttpClientAdapter<HttpClientRequest, HttpClientResponse> {
+	static final class WrappedHttpClientRequest extends brave.http.HttpClientRequest {
+
+		final HttpClientRequest delegate;
+
+		WrappedHttpClientRequest(HttpClientRequest delegate) {
+			this.delegate = delegate;
+		}
+
+		@Override
+		public Object unwrap() {
+			return delegate;
+		}
 
 		@Override
-		public String method(HttpClientRequest request) {
-			return request.method().name();
+		public String method() {
+			return delegate.method().name();
 		}
 
 		@Override
-		public String url(HttpClientRequest request) {
-			return request.uri();
+		public String path() {
+			return delegate.path();
+		}
+
+		@Override
+		public String url() {
+			return delegate.uri();
+		}
+
+		@Override
+		public String header(String name) {
+			return delegate.requestHeaders().get(name);
+		}
+
+		@Override
+		public void header(String name, String value) {
+			delegate.header(name, value);
+		}
+
+	}
+
+	static final class WrappedHttpClientResponse extends brave.http.HttpClientResponse {
+
+		final HttpClientResponse delegate;
+
+		WrappedHttpClientResponse(HttpClientResponse delegate) {
+			this.delegate = delegate;
 		}
 
 		@Override
-		public String requestHeader(HttpClientRequest request, String name) {
-			Object result = request.requestHeaders().get(name);
-			return result != null ? result.toString() : "";
```

Comment:
ps this was not good to return empty strings before. I'm not sure how that practice started...

Reply:
I don't remember, maybe a guard against npe?

Reply:
brave always checks null, so maybe if someone was using this incorrectly perhaps. the impact of "" is that it will invoke a parser for no reason, adding overhead